File: uva/102-ecological-bin-packing/c/main.c
#include <stdio.h>

#define MAX_COLOR 3

#define B 0
#define G 1
#define C 2
#define Moves(x,y,z) curr = b[1][x] + b[2][x] + b[0][y] + b[2][y] + b[0][z] + b[1][z]
#define SelectMin(i) if(curr < min) { min_i = i; min = curr; }

#define Si(i) scanf("%d", &i)

int main() {
  int bin, color, n, curr, min, min_i;
  int b[MAX_COLOR][MAX_COLOR];
  char config_labels[][4] = {
    "BCG",
    "BGC",
    "CBG",
    "CGB",
    "GBC",
    "GCB"
  };

  while(Si(n) != EOF) {
    b[0][0] = n;
    for(color = 1; color < MAX_COLOR; color++) {
      Si(n);
      b[0][color] = n;
    }
    for(bin = 1; bin < MAX_COLOR; bin++)
      for(color = 0; color < MAX_COLOR; color++) {
	Si(n);
	b[bin][color] = n;
      }

    Moves(B, C, G);
    min = curr;
    min_i = 0;
    Moves(B, G, C);
    SelectMin(1);
    Moves(C, B, G);
    SelectMin(2);
    Moves(C, G, B);
    SelectMin(3);
    Moves(G, B, C);
    SelectMin(4);
    Moves(G, C, B);
    SelectMin(5);

    printf("%s %d\n", config_labels[min_i], min);
  }

  return 0;
}

File: uva/102-ecological-bin-packing/python/main.py
#!/usr/bin/env python

import re

'''
There are 3 bins and three kinds of bottles (brown, green, and clear).
Thus, there are 6 candidate configurations.

BCG
BGC
CBG
CGB
GBC
GCB
'''

# ic: initial configuration
# tc: target configuration
def cost(ic, tc):
    cost  = ic[3 + tc[0]] + ic[6 + tc[0]]
    cost += ic[tc[1]]     + ic[6 + tc[1]]
    cost += ic[tc[2]]     + ic[3 + tc[2]]
    return cost

def min(x, y):
    return x if x < y else y

B = 0
G = 1
C = 2
confs = [ [B, C, G],
          [B, G, C],
          [C, B, G],
          [C, G, B],
          [G, B, C],
          [G, C, B] ]
labels = [ 'BCG',
           'BGC',
           'CBG',
           'CGB',
           'GBC',
           'GCB' ]


def main():
    with open('/dev/stdin', 'rt') as f:
        while True:
            line = f.readline().rstrip()
            if not line: break
            # ic matches (B G C | B G C | B G C)
            ic = map(int, re.split(r' +', line))
            # mc: minimum cost
            mc = cost(ic, confs[0])
#            print 'mc0 =', mc
            fc = 0
            for i in range(1, 6):
#                print 'i =', i
#                print 'confs[%d] =' % i, confs[i]
#                print 'cost(ic, confs[%d]) =' % i, cost(ic, confs[i])
                # cc: current cost
                cc = cost(ic, confs[i])
                if cc < mc:
                    mc = cc
                    # fc: final configuration
                    fc = i
#                print 'mc%d =' % i, mc
            print labels[fc], mc

main()
